File: WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m
#import "BlogDetailsViewController.h"

#import "AccountService.h"
#import "BlogService.h"
#import "CommentsViewController.h"
#import "SiteSettingsViewController.h"
#import "SharingViewController.h"
#import "StatsViewController.h"
#import "WPAppAnalytics.h"
#import "WordPress-Swift.h"
#import "MenusViewController.h"

@import Gridicons;
@import Reachability;
@import WordPressData;
@import WordPressShared;

#pragma mark -

@interface BlogDetailsViewController () <UIActionSheetDelegate, UIAlertViewDelegate>

@property (nonatomic, strong) NSArray *headerViewHorizontalConstraints;
@property (nonatomic, strong) BlogService *blogService;

@property (nonatomic) BOOL hasLoggedDomainCreditPromptShownEvent;

@property (nonatomic, strong) BlogDetailsTableViewModel *tableViewModel;

@end

@implementation BlogDetailsViewController

#pragma mark = Lifecycle Methods

- (instancetype)init
{
    self = [super init];

    if (self) {
        self.isScrollEnabled = false;
    }

    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    if (self.isSidebarModeEnabled) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
    } else if (self.isScrollEnabled) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleInsetGrouped];
    } else {
        _tableView = [[IntrinsicTableView alloc] initWithFrame:CGRectZero style:UITableViewStyleInsetGrouped];
        self.tableView.scrollEnabled = false;
    }

    self.tableViewModel = [[BlogDetailsTableViewModel alloc] initWithBlog:self.blog viewController:self];
    [self.tableViewModel configureWithTableView:self.tableView];

    self.tableView.translatesAutoresizingMaskIntoConstraints = false;
    if (self.isSidebarModeEnabled) {
        self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        self.additionalSafeAreaInsets = UIEdgeInsetsMake(0, 8, 0, 0); // Left inset
    }
    [self.view addSubview:self.tableView];
    [self.view pinSubviewToAllEdges:self.tableView];

    UIRefreshControl *refreshControl = [UIRefreshControl new];
    [refreshControl addTarget:self action:@selector(pulledToRefresh) forControlEvents:UIControlEventValueChanged];
    self.tableView.refreshControl = refreshControl;

    self.tableView.accessibilityIdentifier = @"Blog Details Table";

    [WPStyleGuide configureColorsForView:self.view andTableView:self.tableView];
    [WPStyleGuide configureAutomaticHeightRowsFor:self.tableView];

    self.tableView.cellLayoutMarginsFollowReadableWidth = YES;

    self.hasLoggedDomainCreditPromptShownEvent = NO;

    self.blogService = [[BlogService alloc] initWithCoreDataStack:[ContextManager sharedInstance]];
    [self preloadMetadata];

    if (self.blog.account && !self.blog.account.userID) {
        // User's who upgrade may not have a userID recorded.
        AccountService *acctService = [[AccountService alloc] initWithCoreDataStack:[ContextManager sharedInstance]];
        [acctService updateUserDetailsForAccount:self.blog.account success:nil failure:nil];
    }

    [self observeManagedObjectContextObjectsDidChangeNotification];

    [self observeGravatarImageUpdate];

    [self registerForTraitChanges:@[[UITraitHorizontalSizeClass self]] withAction:@selector(handleTraitChanges)];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    [self observeWillEnterForegroundNotification];

    [self.tableViewModel viewWillAppear];

    // Configure and reload table data when appearing to ensure pending comment count is updated
    [self configureTableViewData];

    [self reloadTableViewPreservingSelection];
    [self preloadBlogData];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self createUserActivity];

    [WPAnalytics trackEvent: WPAnalyticsEventMySiteSiteMenuShown];

    if ([self shouldShowJetpackInstallCard]) {
        [WPAnalytics trackEvent:WPAnalyticsEventJetpackInstallFullPluginCardViewed
                     properties:@{WPAppAnalyticsKeyTabSource: @"site_menu"}];
    }

    if ([self shouldShowBlaze]) {
        [ObjCBridge trackBlazeEntryPointDisplayedWithSource:BlazeSourceMenuItem];
    }
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self stopObservingWillEnterForegroundNotification];
}

- (void)handleTraitChanges
{
    // Required to add / remove "Home" section when switching between regular and compact width
    [self configureTableViewData];

    // Required to update disclosure indicators depending on split view status
    [self reloadTableViewPreservingSelection];
}

#pragma mark - Data Model setup

- (void)reloadTableViewPreservingSelection
{
    [self.tableViewModel reloadTableViewPreservingSelection:self.tableView];
}

- (UITableViewScrollPosition)optimumScrollPositionForIndexPath:(NSIndexPath *)indexPath
{
    if (self.isSidebarModeEnabled) {
        return UITableViewScrollPositionNone;
    }
    // Try and avoid scrolling if not necessary
    CGRect cellRect = [self.tableView rectForRowAtIndexPath:indexPath];
    BOOL cellIsNotFullyVisible = !CGRectContainsRect(self.tableView.bounds, cellRect);
    return (cellIsNotFullyVisible) ? UITableViewScrollPositionMiddle : UITableViewScrollPositionNone;
}

- (void)configureTableViewData
{
    [self.tableViewModel configureTableViewData];
}

- (Boolean)isSplitViewDisplayed {
    return self.isSidebarModeEnabled;
}

#pragma mark Site Switching

- (void)switchToBlog:(Blog*)blog
{
    self.blog = blog;
    [self showInitialDetailsForBlog];
    [self.tableView reloadData];
    [self preloadMetadata];
}

- (void)showInitialDetailsForBlog
{
    [self.tableViewModel showInitialDetailsForBlog];
}

#pragma mark - Private methods

- (void)preloadBlogData
{
    // only preload on wifi
    if ([ReachabilityUtils.internetReachability isReachableViaWiFi] == false) {
        return;
    }

    [self preloadComments];
    [self preloadMetadata];
    [self preloadDomains];
}

- (void)preloadComments
{
    CommentService *commentService = [[CommentService alloc] initWithCoreDataStack:[ContextManager sharedInstance]];

    if ([CommentService shouldRefreshCacheFor:self.blog]) {
        [commentService syncCommentsForBlog:self.blog withStatus:CommentStatusFilterAll success:nil failure:nil];
    }
}

- (void)preloadMetadata
{
    __weak __typeof(self) weakSelf = self;
    [self.blogService syncBlogAndAllMetadata:self.blog
                           completionHandler:^{
                               [weakSelf configureTableViewData];
                               [weakSelf reloadTableViewPreservingSelection];
                           }];
}

- (void)preloadDomains
{
    if (![self shouldAddDomainRegistrationRow]) {
        return;
    }

    [self.blogService refreshDomainsFor:self.blog
                                success:nil
                                failure:nil];
}

#pragma mark - Remove Site

- (void)showRemoveSiteAlert
{
    NSString *model = [[UIDevice currentDevice] localizedModel];
    NSString *message = [NSString stringWithFormat:NSLocalizedString(@"Are you sure you want to continue?\n All site data will be removed from your %@.", @"Title for the remove site confirmation alert, %@ will be replaced with iPhone/iPad/iPod Touch"), model];
    NSString *cancelTitle = NSLocalizedString(@"Cancel", nil);
    NSString *destructiveTitle = NSLocalizedString(@"Remove Site", @"Button to remove a site from the app");

    UIAlertControllerStyle alertStyle = [UIDevice isPad] ? UIAlertControllerStyleAlert : UIAlertControllerStyleActionSheet;
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil
                                                                             message:message
                                                                      preferredStyle:alertStyle];

    [alertController addCancelActionWithTitle:cancelTitle handler:nil];
    [alertController addDestructiveActionWithTitle:destructiveTitle handler:^(UIAlertAction * __unused action) {
        [self confirmRemoveSite];
    }];

    [self presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Notification handlers

- (void)handleDataModelChange:(NSNotification *)note
{
    NSSet *deletedObjects = note.userInfo[NSDeletedObjectsKey];
    if ([deletedObjects containsObject:self.blog]) {
        [self.navigationController popToRootViewControllerAnimated:NO];
        return;
    }

    if (self.blog.account == nil || self.blog.account.isDeleted) {
        // No need to reload this screen if the blog's account is deleted (i.e. during logout)
        return;
    }

    NSSet *updatedObjects = note.userInfo[NSUpdatedObjectsKey];
    if ([updatedObjects containsObject:self.blog] || [updatedObjects containsObject:self.blog.settings]) {
        [self configureTableViewData];
        [self reloadTableViewPreservingSelection];
    }
}

- (void)handleWillEnterForegroundNotification:(NSNotification *)note
{
    [self configureTableViewData];
    [self reloadTableViewPreservingSelection];
}

- (void)observeManagedObjectContextObjectsDidChangeNotification
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleDataModelChange:)
                                                 name:NSManagedObjectContextObjectsDidChangeNotification
                                               object:context];
}

- (void)observeWillEnterForegroundNotification
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleWillEnterForegroundNotification:)
                                                 name:UIApplicationWillEnterForegroundNotification
                                               object:nil];
}

- (void)stopObservingWillEnterForegroundNotification
{
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationWillEnterForegroundNotification
                                                  object:nil];
}

#pragma mark - UIViewControllerTransitioningDelegate

- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented presentingViewController:(UIViewController *)presenting sourceViewController:(UIViewController *)source
{
    if ([presented isKindOfClass:[FancyAlertViewController class]]) {
        return [[FancyAlertPresentationController alloc] initWithPresentedViewController:presented
                                                                presentingViewController:presenting];
    }

    return nil;
}

#pragma mark - UIAdaptivePresentationControllerDelegate

- (void)presentationControllerWillDismiss:(UIPresentationController *)presentationController {
    if (presentationController.presentedViewController == self.presentedSiteSettingsViewController) {
        [self.tableView deselectSelectedRowWithAnimation:YES];
    }
}

#pragma mark - Domain Registration

- (void)updateTableViewAndHeader
{
    [self updateTableView:^{}];
}

/// This method syncs the blog and its metadata, then reloads the table view.
///
- (void)updateTableView:(void(^)(void))completion
{
    __weak __typeof(self) weakSelf = self;
    [self.blogService syncBlogAndAllMetadata:self.blog
                           completionHandler:
     ^{
        [weakSelf configureTableViewData];
        [weakSelf reloadTableViewPreservingSelection];
        completion();
    }];
}

#pragma mark - Pull To Refresh

- (void)pulledToRefresh {
    [self pulledToRefreshWith:self.tableView.refreshControl onCompletion:^{}];
}

- (void)pulledToRefreshWith:(UIRefreshControl *)refreshControl onCompletion:( void(^)(void))completion {

    [self updateTableView: ^{
        // WORKAROUND: if we don't dispatch this asynchronously, the refresh end animation is clunky.
        // To recognize if we can remove this, simply remove the dispatch_async call and test pulling
        // down to refresh the site.
        dispatch_async(dispatch_get_main_queue(), ^(void){
            [refreshControl endRefreshing];

            completion();
        });
    }];
}

@end

File: WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.swift
import UIKit
import SwiftUI
import WordPressData
import WordPressShared
import WordPressUI
import Gridicons

public protocol BlogDetailsPresentationDelegate: AnyObject {
    func presentBlogDetailsViewController(_ viewController: UIViewController)
}

public class BlogDetailsViewController: UIViewController {

    public var blog: Blog {
        didSet {
            tableViewModel?.blog = blog
        }
    }
    public private(set) var tableViewModel: BlogDetailsTableViewModel?
    public weak var presentationDelegate: BlogDetailsPresentationDelegate?
    public var isSidebarModeEnabled = false
    public weak var presentedSiteSettingsViewController: UIViewController?

    var headerViewController: UIViewController? {
        didSet { hostingController?.rootView.headerViewController = headerViewController }
    }
    lazy var onRefresh: () async -> Void = { [weak self] in
        await self?.refreshMenu()
    }

    private var hostingController: UIHostingController<BlogDetailsView>?

    private lazy var blogService = BlogService(coreDataStack: ContextManager.shared)
    private var hasLoggedDomainCreditPromptShownEvent = false

    init(blog: Blog) {
        self.blog = blog
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override public func viewDidLoad() {
        super.viewDidLoad()

        let viewModel = BlogDetailsTableViewModel(blog: blog, viewController: self)
        tableViewModel = viewModel
        // Deep links need the menu actions before the view appears.
        viewModel.configureTableViewData()

        let host = UIHostingController(
            rootView: BlogDetailsView(
                viewModel: viewModel,
                headerViewController: headerViewController,
                refresh: { [weak self] in await self?.onRefresh() }
            )
        )
        hostingController = host
        host.view.backgroundColor = .clear
        add(host)
        host.view.translatesAutoresizingMaskIntoConstraints = false
        view.pinSubviewToAllEdges(host.view)

        view.backgroundColor = isSidebarModeEnabled ? .systemBackground : .systemGroupedBackground

        hasLoggedDomainCreditPromptShownEvent = false
        preloadMetadata()

        if let account = blog.account, account.userID == nil {
            let service = AccountService(coreDataStack: ContextManager.shared)
            service.updateUserDetails(for: account, success: nil, failure: nil)
        }

        observeManagedObjectContextObjectsDidChangeNotification()
        observeGravatarImageUpdate()
        downloadGravatarImage()

        registerForTraitChanges([UITraitHorizontalSizeClass.self], action: #selector(handleTraitChanges))
    }

    override public func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        observeWillEnterForegroundNotification()
        tableViewModel?.viewWillAppear()
        // Configure and reload table data when appearing to ensure pending comment count is updated
        configureTableViewData()
        preloadBlogData()
    }

    override public func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        createUserActivity()

        WPAnalytics.track(
            .mySiteSiteMenuShown,
            properties: [
                "has_application_password": (try? blog.getApplicationToken()) != nil ? "true" : "false"
            ],
            blog: blog
        )

        if shouldShowJetpackInstallCard() {
            WPAnalytics.track(.jetpackInstallFullPluginCardViewed, properties: [WPAppAnalyticsKeyTabSource: "site_menu"])
        }

        if shouldShowBlaze() {
            BlazeEventsTracker.trackEntryPointDisplayed(for: .menuItem, blogProperties: blog.analyticsProperties)
        }
    }

    override public func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        stopObservingWillEnterForegroundNotification()
    }

    @objc private func handleTraitChanges() {
        configureTableViewData()
    }

    public func configureTableViewData() {
        tableViewModel?.configureTableViewData()
    }

    public func `switch`(to blog: Blog) {
        self.blog = blog
        showInitialDetailsForBlog()
        preloadMetadata()
    }

    public func showInitialDetailsForBlog() {
        tableViewModel?.showInitialDetailsForBlog()
    }

    @MainActor
    func refreshMenu() async {
        await withCheckedContinuation { continuation in
            blogService.syncBlogAndAllMetadata(blog) {
                continuation.resume()
            }
        }

        if let service = CustomPostTypeService(blog: blog) {
            tableViewModel?.hasCustomPostTypes = (try? await service.customTypes())?.isEmpty == false
        } else {
            tableViewModel?.hasCustomPostTypes = false
        }

        configureTableViewData()
    }

    public func refresh() {
        Task { await refreshMenu() }
    }

    private func preloadBlogData() {
        // only preload on wifi
        guard ReachabilityUtils.isReachableViaWiFi() else {
            return
        }

        preloadComments()
        preloadMetadata()
        preloadDomains()
    }

    private func preloadComments() {
        let commentService = CommentService(coreDataStack: ContextManager.shared)

        if CommentService.shouldRefreshCache(for: blog) {
            commentService.syncComments(for: blog, withStatus: CommentStatusFilterAll, success: nil, failure: nil)
        }
    }

    public func preloadMetadata() {
        Task {
            await refreshMenu()
        }
    }

    private func preloadDomains() {
        guard shouldAddDomainRegistrationRow() else {
            return
        }

        blogService.refreshDomains(for: blog, success: nil, failure: nil)
    }

    public func showRemoveSiteAlert() {
        let model = UIDevice.current.localizedModel
        let message = String(format: NSLocalizedString(
            "Are you sure you want to continue?\n All site data will be removed from your %@.",
            comment: "Title for the remove site confirmation alert, %@ will be replaced with iPhone/iPad/iPod Touch"
        ), model)

        let destructiveTitle = NSLocalizedString("Remove Site", comment: "Button to remove a site from the app")

        let alertStyle: UIAlertController.Style = UIDevice.isPad() ? .alert : .actionSheet
        let alertController = UIAlertController(title: nil, message: message, preferredStyle: alertStyle)

        alertController.addCancelActionWithTitle(SharedStrings.Button.cancel, handler: nil)
        alertController.addDestructiveActionWithTitle(destructiveTitle) { [weak self] _ in
            self?.confirmRemoveSite()
        }

        present(alertController, animated: true)
    }

    @objc private func handleDataModelChange(_ notification: NSNotification) {
        guard let deletedObjects = notification.userInfo?[NSDeletedObjectsKey] as? Set<NSManagedObject> else {
            return
        }

        if deletedObjects.contains(blog) {
            navigationController?.popToRootViewController(animated: false)
            return
        }

        if blog.account == nil || blog.account?.isDeleted == true {
            return
        }

        guard let updatedObjects = notification.userInfo?[NSUpdatedObjectsKey] as? Set<NSManagedObject> else {
            return
        }

        if updatedObjects.contains(blog) || (blog.settings != nil && updatedObjects.contains(blog.settings!)) {
            configureTableViewData()
        }
    }

    @objc private func handleWillEnterForeground(_ notification: NSNotification) {
        configureTableViewData()
    }

    private func observeManagedObjectContextObjectsDidChangeNotification() {
        let context = ContextManager.shared.mainContext
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleDataModelChange(_:)),
            name: NSNotification.Name.NSManagedObjectContextObjectsDidChange,
            object: context
        )
    }

    private func observeWillEnterForegroundNotification() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleWillEnterForeground(_:)),
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )
    }

    private func stopObservingWillEnterForegroundNotification() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )
    }
}

extension BlogDetailsViewController: UIViewControllerTransitioningDelegate {

    public func presentationController(
        forPresented presented: UIViewController,
        presenting: UIViewController?,
        source: UIViewController
    ) -> UIPresentationController? {
        if presented is FancyAlertViewController {
            return FancyAlertPresentationController(
                presentedViewController: presented,
                presenting: presenting
            )
        }
        return nil
    }
}

extension BlogDetailsViewController: UIAdaptivePresentationControllerDelegate {

    public func presentationControllerWillDismiss(_ presentationController: UIPresentationController) {
        if presentationController.presentedViewController == presentedSiteSettingsViewController {
            tableViewModel?.clearSelection()
        }
    }
}
